### PR TITLE
Add ubuntu22 artifacts test

### PR DESCRIPTION
**Pull request by KnifeyMoloko**
Adding artifacts tests for ubuntu22.04. 

Jenkins runs: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/artifacts-ubuntu/
Issue open in scylla repo for `scylla_setup` error: https://github.com/scylladb/scylla/issues/10669
Trello task: https://trello.com/c/ChdoRZBV

Sanity runs with other distros:
Ubuntu 20.04 -> https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/artifacts-2004/4/ [PASSED]
Centos8 -> https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/artfiacts-centos-8/2/ [PASSED]
Rocky8 -> https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/artifacts-rocky8/2/console [PASSED]
Oel81 -> https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/artifacts-oel81/3/ [PASSED]
Ubuntu 20.04 ARM -> https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/artifacts-2004-arm/ [PASSED]
Ubuntu 22.04 ARM -> https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/artifacts-2204-arm/ [FAILED]

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

**Review by fgelcer — COMMENTED** [review]
LGTM

LMK when it will be ready to go, and please, add labels to backport (to 5.0 and 2022.1)

**Comment by fgelcer** [comment]
also, when you have it, please add a link to a running job

**Comment by roydahan** [comment]
Why is it still on draft?

**Comment by roydahan** [comment]
@KnifeyMoloko / @fgelcer can we push this forward?
It needs to reach 5.0 and 2022.1

File: configurations/arm/ubuntu2204.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+ami_db_scylla_user: 'ubuntu'
+ami_id_db_scylla: 'ami-0127b718f588563a0'
```
Comment (fruch) [inline configurations/arm/ubuntu2204.yaml:2]: Can you add the name of the ami in the comments?

File: sdcm/cluster_gce.py [hunk 2]
```diff
@@ -319,6 +319,12 @@ def set_tags_as_labels(_instance):
                 sudo systemctl disable sshguard
                 sudo systemctl stop sshguard
             """)
+
+        if self.params.get("scylla_linux_distro") in ("ubuntu-jammy"):
```
Comment (fruch) [inline sdcm/cluster_gce.py:323]: Can you check if this has an effect on other distros ?
As far as I know, those parameters aren't new.

Maybe we could just set it across the board for all ?

File: sdcm/cluster_gce.py [hunk 3]
```diff
@@ -366,14 +373,17 @@ def _create_node_with_retries(self,
                                   create_node_params: dict[str, Any],
                                   spot: bool = False) -> GCENode:
         try:
-            return self._gce_services[dc_idx].create_node(**create_node_params)
+            response = self._gce_services[dc_idx].create_node(**create_node_params)
+            self.log.info(f"Got GCE create node response:\n{response}")
```
Comment (fruch) [inline sdcm/cluster_gce.py:377]: please put this print in debug

File: sdcm/cluster_gce.py (hunk 2 above)
Comment (fruch) [reply sdcm/cluster_gce.py:323]: Also it might be safer to use the distrio.UBUNTU22

Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py]: I decided to remove this and add the specific script lines to `get_startup_script` in `TestConfig`, as the cluster logic is already confusing as it is.

File: sdcm/cluster_gce.py (hunk 3 above)
Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py:372]: I'll remove this, as it's just a debug log for my needs

File: test-cases/artifacts/ubuntu2204.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+gce_n_local_ssd_disk_db: 1
+instance_provision: "spot"
+instance_provision_fallback_on_demand: true
+logs_transport: 'ssh'
+n_db_nodes: 1
+n_loaders: 0
+n_monitor_nodes: 0
+nemesis_class_name: 'NoOpMonkey'
+scylla_linux_distro: 'ubuntu-jammy'
+scylla_repo: 'http://downloads.scylladb.com/unstable/scylla/master/deb/unified/latest/scylladb-master/scylla.list'
```
[the remark this answers is not in the document]
Comment (KnifeyMoloko) [reply test-cases/artifacts/ubuntu2204.yaml:16]: Tbh this just mirrors what we have for other ubuntu artifact test cases. But it does point to the repo that *should* have the latest master

File: sdcm/cluster_gce.py (hunk 3 above)
Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py]: Fixed (removed)

File: configurations/arm/ubuntu2204.yaml (hunk 1 above)
Comment (KnifeyMoloko) [reply configurations/arm/ubuntu2204.yaml:2]: I can't find this ami for some reason... I'm thinking if we even need this. We're providing the `scylla_repo` param in artifacts tests and these 2 should not be used together either way, IIUC. @roydahan Are these values in `/configurations` necessary?

**Comment by KnifeyMoloko** [comment]
> is there a running job? is it already passing some tests?

@fgelcer Yes and no, respectively. The job is running, but `scylla_setup` is failing on them. I'll create a new issue for it tomorrow.

File: sdcm/provision/common/configuration_script.py [hunk 5]
```diff
@@ -32,10 +32,12 @@ class ConfigurationScriptBuilder(AttrBuilder, metaclass=abc.ABCMeta):
     configure_sshd: bool = True
     disable_ssh_while_running: bool = False
     hostname: str = ''
+    legacy_rsa_keys: bool = True
 
     def to_string(self) -> str:
         script = self._start_script()
         script += self._script_body()
+        script += self._enable_legacy_rsa_keys()
```
Comment (fgelcer) [inline sdcm/provision/common/configuration_script.py:40]: should this be called only for Ubuntu22? or it doesn't matter for other versions (including RHEL, Debian and old Ubuntu ones)?

File: sdcm/test_config.py [hunk 6]
```diff
@@ -252,6 +252,8 @@ def get_startup_script(cls) -> str:
             syslog_host_port=host_port,
             logs_transport=cls._tester_obj.params.get('logs_transport') if cls._tester_obj else "rsyslog",
             disable_ssh_while_running=True,
+            configure_sshd=True,
+            legacy_rsa_keys=True
```
Comment (fgelcer) [inline sdcm/test_config.py:256]: again, would this be called to all tests? or should be done for Ubuntu22 only?

File: test-cases/artifacts/ubuntu2204.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgelcer) [reply test-cases/artifacts/ubuntu2204.yaml:16]: what we have in this parameter, for release branches? also pointing to latest master?

**Review by fgelcer — COMMENTED** [review]
little questions, but other than that, LGTM

File: configurations/arm/ubuntu2204.yaml (hunk 1 above)
Comment (fruch) [reply configurations/arm/ubuntu2204.yaml:2]: since we are running in AWS, we need this as the base image, regardless if we using `scylla_repo` to install or not.
this image is `ubuntu/images/hvm-ssd/ubuntu-focal-20.04-arm64-server-20200701` (it's public image)
but it's not 22.04 as this PR titled.

you probably want to use this:
```yaml
ami_id_db_scylla: ami-022c8ce295ce9ac4c  # ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-arm64-server-20220506 
```

File: sdcm/provision/common/configuration_script.py [hunk 7]
```diff
@@ -89,6 +91,15 @@ def _script_body(self) -> str:
             script += 'systemctl start sshd || true\n'
         return script
 
+    @staticmethod
+    def _enable_legacy_rsa_keys():
+        """
+        Add support for RSA-1 ssh keys in sshd config.
+        """
+        script = configure_legacy_rsa_keys_for_sshd()
+        script += restart_sshd_service()
```
Comment (fruch) [inline sdcm/provision/common/configuration_script.py:100]: why not adding the whole thing into `configure_sshd_script` ? why do we need multiple restarts of the sshd ?

File: configurations/arm/ubuntu2204.yaml (hunk 1 above)
Comment (KnifeyMoloko) [reply configurations/arm/ubuntu2204.yaml]: Ah, that's correct. I added the base image and comment.

File: test-cases/artifacts/ubuntu2204.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (KnifeyMoloko) [reply test-cases/artifacts/ubuntu2204.yaml:16]: Yup, it also points to master in branch-4.6 for example. We provide the actual param in the job params though, so that gets used instead of the test case config value.

File: sdcm/provision/common/configuration_script.py (hunk 5 above)
Comment (KnifeyMoloko) [reply sdcm/provision/common/configuration_script.py:40]: It shouldn't interfere with other versions, I think. It's necessary for any distro that uses the more recent version of openssh, so having this here should cover all such cases.

Comment (fruch) [reply sdcm/provision/common/configuration_script.py:40]: did you test some other version or distros ?

File: sdcm/test_config.py (hunk 6 above)
Comment (KnifeyMoloko) [reply sdcm/test_config.py:256]: IMHO this can be called for all tests.

File: sdcm/provision/common/configuration_script.py (hunk 5 above)
Comment (KnifeyMoloko) [reply sdcm/provision/common/configuration_script.py:40]: Would be happy to do so, waiting for available executors on Jenkins currently.

File: configurations/arm/ubuntu2204.yaml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+ami_db_scylla_user: 'ubuntu'
+ami_id_db_scylla: 'ami-022c8ce295ce9ac4c'  # ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-arm64-server-20220506
+region_name: 'eu-west-1'
+instance_type_db: 'r6gd.large'
```
Comment (roydahan) [inline configurations/arm/ubuntu2204.yaml:4]: Why do we use this instance type?
It's not the recommended instance type for scylla.
If we do it in other OS versions, please change that as well.
Please use: 
**im4gn.xlarge**

File: test-cases/artifacts/ubuntu2204.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (roydahan) [reply test-cases/artifacts/ubuntu2204.yaml:16]: @fruch is right, please add a comment that this is something that is being override by the Jenkins job parameter.

File: test-cases/artifacts/ubuntu2204.yaml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+scylla_apt_keys:
+  # When will use manager 2.6, need to remove this commit because it's a workaround to make artifacts works
+  - '5E08FBD8B5D6EC9C'  # ScyllaDB Package Signing Key 2020 <security@scylladb.com>
```
Comment (roydahan) [inline test-cases/artifacts/ubuntu2204.yaml:19]: I think we don't need this one, since it's for the older versions.
But please test.

File: sdcm/provision/common/utils.py [hunk 10]
```diff
@@ -120,6 +120,14 @@ def configure_sshd_script():
     """)
 
 
+def configure_legacy_rsa_keys_for_sshd():
```
Comment (roydahan) [inline sdcm/provision/common/utils.py:123]: This commit doesn't have any description and I couldn't understand why it was needed without reading the comments here.
Please explain the requirement and the fact that it affects all distros.

In addition, the title is wrong, it shouldn't be "fix(nemesis)".

**Review by roydahan — CHANGES_REQUESTED** [review]

**Comment by enaydanov** [comment]
@KnifeyMoloko please, don't miss unit tests for `Distro.is_ubuntu22`

File: configurations/arm/ubuntu2204.yaml (hunk 8 above)
Comment (KnifeyMoloko) [reply configurations/arm/ubuntu2204.yaml]: Fixed

File: configurations/arm/ubuntu2204.yaml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+ami_db_scylla_user: 'ubuntu'
+ami_id_db_scylla: 'ami-022c8ce295ce9ac4c'  # ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-arm64-server-20220506
+region_name: 'eu-west-1'
+instance_type_db: 'im4gn.xlarge'
```
Comment (roydahan) [inline configurations/arm/ubuntu2204.yaml:4]: @KnifeyMoloko please send another PR to change that also in all other arm artifact tests.

File: test-cases/artifacts/ubuntu2204.yaml (hunk 9 above)
Comment (roydahan) [reply test-cases/artifacts/ubuntu2204.yaml]: @KnifeyMoloko did you remove both keys?
We must need at least one of them.

**Comment by roydahan** [comment]
@KnifeyMoloko please add links to the staging jobs, when they pass we need to copy them to master and branch-5.0 and 2022.1.

File: test-cases/artifacts/ubuntu2204.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (KnifeyMoloko) [reply test-cases/artifacts/ubuntu2204.yaml]: @roydahan I removed the line completely. I agree with @fruch that it would be better to fail fast if somehow the job doesn't provide the value. My runs work ok without this line.

File: sdcm/provision/common/configuration_script.py (hunk 5 above)
Comment (KnifeyMoloko) [reply sdcm/provision/common/configuration_script.py]: Tested on other distros (previous ubuntu versions) and got some ssh failures, so I made it so that this line will only be called with Ubuntu22.04 for now.

File: test-cases/artifacts/ubuntu2204.yaml (hunk 9 above)
Comment (KnifeyMoloko) [reply test-cases/artifacts/ubuntu2204.yaml]: Sorry, a bit to eager with the removals. I left the 2022 Signing Key now.

**Comment by KnifeyMoloko** [comment]
> @KnifeyMoloko please, don't miss unit tests for `Distro.is_ubuntu22` (and the property too :)

Yup :) I added both the property and distro unit tests.

File: configurations/arm/ubuntu2204.yaml (hunk 11 above)
Comment (KnifeyMoloko) [reply configurations/arm/ubuntu2204.yaml:4]: Sure thing: https://github.com/scylladb/scylla-cluster-tests/pull/4811

File: sdcm/cluster_gce.py [hunk 12]
```diff
@@ -311,14 +311,17 @@ def set_tags_as_labels(_instance):
         # Name must start with a lowercase letter followed by up to 63
         # lowercase letters, numbers, or hyphens, and cannot end with a hyphen
         assert len(name) <= 63, "Max length of instance name is 63"
-        startup_script = self.test_config.get_startup_script()
+        startup_script = self.test_config.get_startup_script(
```
Comment (fruch) [inline sdcm/cluster_gce.py:314]: get_startup_script is being called from multiple places
Did you run the ARM tests? They are running in AWS, seems it's not currently handled

Also when I think about it, we might need to do the figuring out inside the startup script, casue what would happen when our AMI would move to that Ubuntu version (also any older distro might get ssh update, regardless)

Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py]: Agreed about moving the check inside the script. It's condition is now for ssh version to be greater than 8.8.

As for the ARM tests: the test for the previous Ubuntu release (20.04) work without any issues. Ubuntu 22.04 however bumps into the same issue as the regular GCE test had - ssh-rsa key format not being supported. It looks like the user-data script that resolves the issue on GCE is not being executed at all on AWS. Therefore we can't modify `sshd_config` and the job fails while waiting for ssh to be up.

Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py]: @roydahan @fruch I created a new issue for our config script: https://github.com/scylladb/scylla-cluster-tests/issues/4818 This is also related to a `scylla-machine-image` issue Łukasz brought up: https://github.com/scylladb/scylla-machine-image/issues/357 In short, since we're not taking advantage of `cloud-init` and delegating config operations to `scylla-cloud-image`, we're opening ourselves to issues like the one with `openssh`. This particular one can be solved by changing our ssh keys, but in general it would be great to make use of `cloud-init`.

Comment (fruch) [reply sdcm/cluster_gce.py]: the bug that SCT breaks the support of cloud-init, it's directly related to scylla-machine-image. it would be nice that it would be the same, but we need to address it SCT regardless of a change in machine-image.

**Comment by fruch** [comment]
@fgelcer we still have an issue with non artifacts tests running in AWS. which this PR doesn't solve.

**Review by fruch — DISMISSED** [review]
LGTM, but we need to open an issue/PR for handling the arm test for it.

**Comment by fgelcer** [comment]
> LGTM, but we need to open an issue/PR for handling the arm test for it.

where is the fix supposed to be done? SMI?

**Comment by fruch** [comment]
> > LGTM, but we need to open an issue/PR for handling the arm test for it.
> 
> where is the fix supposed to be done? SMI?

SCT needs to support both type of user-data, cloud-int and SMI. 

later SMI can support both, and SCT can align to that.

since this is need for 5.0 / 2022.1, such a change in SMI won't get into those branches now.

**Comment by fruch** [comment]
@KnifeyMoloko please create the relevent jobs on master for this PR.

also open a PR for scylla-pkg to start triggering those artifacts jobs

**Comment by fgelcer** [comment]
> @KnifeyMoloko please create the relevent jobs on master for this PR.
> 
> also open a PR for scylla-pkg to start triggering those artifacts jobs

they already exist for master (just the offline install ones are missing)
for 5.0 only the offline install ones are missing
for 2022.1 both regular artifacts and offline are missing.

once that is done, scylla-pkg should start triggering them all

**Comment by KnifeyMoloko** [comment]
> > @KnifeyMoloko please create the relevent jobs on master for this PR.
> > also open a PR for scylla-pkg to start triggering those artifacts jobs
> 
> they already exist for master (just the offline install ones are missing) for 5.0 only the offline install ones are missing for 2022.1 both regular artifacts and offline are missing.
> 
> once that is done, scylla-pkg should start triggering them all

All relevant jobs are created now both for regular and offline-installer. https://github.com/scylladb/scylla-pkg/pull/2935 is the scylla-pkg PR for the triggers

**Comment by KnifeyMoloko** [comment]
@scylladb/qa-maintainers This should be ok. to backport to 5.0. and 2022.1 now.
